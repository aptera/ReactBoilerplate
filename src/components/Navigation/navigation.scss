.nav-react
{
    padding: 0;
    margin: 0;
    list-style: none;

    li
    {
        a
        {
            display: block;
            font-size: 18px;
            padding: 5px;
            border: solid 1px black;
            margin: 2px 0;
            text-decoration: none;
            color: #000;
            background-color: hotpink;
            transition: all 250ms;
            &:hover
            {
                background-color: lighten(gray, 20%);
                border-radius: 5px;
            }
        }
        ul {
            list-style: none;
        }
    }
}
