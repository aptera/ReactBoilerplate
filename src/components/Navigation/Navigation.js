import React, { Component } from 'react';
import values from 'object.values';
import './navigation.scss';

class Navigation extends Component {

    constructor(props) {
        super(props);
        this.props = props;
        this.state = {
            model: props.Model
        };
    }
    
    renderRootLevelNode(node) {
        return (
            <li key={'root_' + node.Title}>
                <a href={node.Url}>{node.Title}</a>
                {this.renderSubLevelsRecursive(node, 0)}
            </li>
        );
    }

    renderSubLevelsRecursive(node, level)
    {
        if (!node.ChildNodes || node.ChildNodes.length == 0)
            return;

        return (
            <ul>
                {values(node.ChildNodes).map((node) =>
                    <li key={'sub_' + node.Title}>
                        <a href={node.Url}>{node.Title}</a>
                        {this.renderSubLevelsRecursive(node, level + 1)}
                    </li>
                )}
            </ul>
        );
    }

    render() {
        return (
            <div className={this.props.Model.CssClass}>
                <ul className="nav nav-react">
                    {values(this.props.Model.Nodes).map((node) =>
                        this.renderRootLevelNode(node)
                    )}
                </ul>
            </div>
        );
    }
}

export default Navigation;