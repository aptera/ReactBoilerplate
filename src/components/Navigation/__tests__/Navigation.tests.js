import React from 'react';

import renderer from 'react-test-renderer';
import ReactTestUtils from 'react-dom/test-utils';

// import the component to put under test.
import Navigation from '../Navigation';

describe('the navigation component', () => {

    it('renders correctly', () => {
        let model = {
            CssClass: 'testcss',
            Nodes: [
                {
                    Title: 'Root Level Node 0',
                    Url: 'http://google.com',
                    ChildNodes: []
                },
                {
                    Title: 'Root Level Node 1',
                    Url: 'http://google.com',
                    ChildNodes: [
                        {
                            Title: 'Sub Level 1-0',
                            Url: 'http://google.com',
                            ChildNodes: []
                        },
                        {
                            Title: 'Sub Level 1-1',
                            Url: 'http://google.com',
                            ChildNodes: [
                                {
                                    Title: 'Sub Level 1-1-0',
                                    Url: 'http://google.com',
                                    ChildNodes: []
                                },
                            ]
                        }
                    ]
                },
                {
                    Title: 'Root Level Node 2',
                    Url: 'http://google.com',
                    ChildNodes: [
                        {
                            Title: 'Sub Level 2-0',
                            Url: 'http://google.com'
                        }
                    ]
                }
            ]
        }
        const tree = renderer.create(
            <Navigation Model={model} />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});