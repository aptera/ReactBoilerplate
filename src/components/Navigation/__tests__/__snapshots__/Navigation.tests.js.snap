// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the navigation component renders correctly 1`] = `
<div
  className="testcss"
>
  <ul
    className="nav nav-react"
  >
    <li>
      <a
        href="http://google.com"
      >
        Root Level Node 0
      </a>
    </li>
    <li>
      <a
        href="http://google.com"
      >
        Root Level Node 1
      </a>
      <ul>
        <li>
          <a
            href="http://google.com"
          >
            Sub Level 1-0
          </a>
        </li>
        <li>
          <a
            href="http://google.com"
          >
            Sub Level 1-1
          </a>
          <ul>
            <li>
              <a
                href="http://google.com"
              >
                Sub Level 1-1-0
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </li>
    <li>
      <a
        href="http://google.com"
      >
        Root Level Node 2
      </a>
      <ul>
        <li>
          <a
            href="http://google.com"
          >
            Sub Level 2-0
          </a>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
